internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +115,173 @@ func New(ctx context.Context,
+		upstreamResponseChan := o.upstreamClient.OpenStream(ctx, &req)
+		respChannel := o.upstreamResponseMap.add(aggregatedKey, upstreamResponseChan)
+		// Spin up a go routine to watch for upstream responses.
+		// One routine is opened per aggregate key.
+		go o.watchUpstream(ctx, aggregatedKey, respChannel.response, respChannel.done)
+
+	}
+	o.upstreamResponseMap.mu.Unlock()
+
+	return responseChannel, o.onCancelWatch(&req)
 }
 
 // Fetch implements the polling method of the config cache using a non-empty request.
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *upstream.Response,
+	done <-chan bool,
+) {
+	for {
+		select {
+		case x := <-responseChannel:
+			if x.Err != nil {
+				// A problem occurred fetching the response upstream, log and
+				// return the most recent cached response, so that the
+				// downstream will reissue the discovery request.
+				o.logger.With("err", x.Err).With("key", aggregatedKey).Error(ctx, "upstream error")
+			} else {
+				// Cache the response.
+				_, err := o.cache.SetResponse(aggregatedKey, x.Response)
+				if err != nil {
+					// If we fail to cache the new response, log and return the old one.
+					o.logger.With("err", err).With("key", aggregatedKey).
+						With("resp", x.Response).Error(ctx, "Failed to cache the response")
+				}
+			}
+
+			// Get downstream watches and fan out.
+			// We retrieve from cache rather than directly fanning out the
+			// newly received response because the cache does additional
+			// resource serialization.
+			cached, err := o.cache.Fetch(aggregatedKey)
+			if err != nil {
+				o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "cache fetch failed")
+				// Can't do anything because we don't know who the watches
+				// are. Drop the response.
+			} else {
+				if cached == nil || cached.Resp == nil {
+					// If cache is empty, there is nothing to fan out.
+					if x.Err != nil {
+						// Warn. Benefit of the doubt that this is the first request.
+						o.logger.With("key", aggregatedKey).
+							Warn(ctx, "attempted to fan out with no cached response")
+					} else {
+						// Error. Sanity check. Shouldn't ever reach this.
+						o.logger.With("key", aggregatedKey).
+							Error(ctx, "attempted to fan out with no cached response")
+					}
+				} else {
+					// Goldenpath.
+					o.fanout(cached.Resp, cached.Requests)
+				}
+			}
+		case <-done:
+			// Exit when signaled that the stream has closed.
```

Comment:
A stream should never be closed. The reason a stream was open was because a sidecar asked for a resource. If the key cannot be found on upstream, we should just retry(with backoff) .

Reply:
I'm using this during shutdown and testing, please see the updated commits.